python/cudf_polars/cudf_polars/dsl/utils/windows.py

```diff
@@ -140,14 +155,23 @@ def offsets_to_windows(
     Returns
     -------
     tuple of preceding and following windows as pylibcudf scalars.
+
+    Notes
+    -----
+    The returned scalars are constructed on a new stream, which is
+    synchronized before being returned.
     """
     offset_i = duration_to_int(dtype, *offset)
     period_i = duration_to_int(dtype, *period)
     # Polars uses current_row + offset, ..., current_row + offset + period
     # Libcudf uses current_row - preceding, ..., current_row + following
-    return duration_to_scalar(dtype, -offset_i), duration_to_scalar(
-        dtype, offset_i + period_i
+    stream = get_cuda_stream()
+    result = (
+        duration_to_scalar(dtype, -offset_i, stream=stream),
+        duration_to_scalar(dtype, offset_i + period_i, stream=stream),
     )
+    stream.synchronize()
```

Comment:
general question: Is it an issue if an exception is raised before the stream is synchronized?

Reply:
My guess (just a guess) is that whether or not that's an issue is independent of whether an exception is raised, and has more to do with how the exception is handled. Like, this snippet that swallows the exception preventing `stream.syncrhonize()` would be unsafe:


```
stream = Stream()
stream_ordered_data = operation_on_data(..., stream=stream)

try:
    1/0
    stream.synchronize()
except Exception:
    pass

do_some_operation_requiring_syncrhonized_data(stream_ordered_data)
```

in the same way that just forgetting to call `stream.synchronize` would be unsafe. But that's a property of just not reaching the `stream.synchronize()`.

Does that address your question?